<a href="https://colab.research.google.com/github/ZeroAthena/SimMat02025/blob/main/Tarea_1_3_Programaci%C3%B3n_lineal.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables:

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

In [1]:
import scipy.optimize as opt
import numpy as np
import matplotlib.pyplot as plt
import sympy as sym

In [12]:
#Restricciones:
#Costo - x1 + x2 + x3 + x4 + x5 + x6 = 100
#Proteína - 0.1x1 + 0.2x2 + 0.15x3 + 0x4 + 0.04x5 +0x6 ≥ 8
#Grasa - 0.08x1 + 0.1x2 + 0.11x3 + 0.01x4 + 0.01x5 + 0x6 ≥ 6
#Fibra - 0.001x1 + 0.005x2 + 0.005x3 + 0.1x4 + 0.15x5 + 0x6 ≤ 2
#Sal - 0.002x1 + 0.005x2 + 0.007x3 + 0.002x4 + 0.008x5 + 0x6 ≤ 0.4

# Costos
c = np.array([0.013,0.008,0.010,0.002,0.005,0.001])

# Restricciones
a = np.array([
    [-0.1,-0.2,-0.15,0,-0.04,0],      # proteína ≥8  ->  -prot ≤ -8
    [-0.08,-0.1,-0.11,-0.01,-0.01,0], # grasa ≥6     ->  -grasa ≤ -6
    [0.001,0.005,0.005,0.1,0.15,0],   # fibra ≤2
    [0.002,0.005,0.007,0.002,0.008,0]]) # sal ≤0.4

b = np.array([-8,-6,2,0.4])

# Restricción = 100 g
d = np.array([[1,1,1,1,1,1]])
e = np.array([100])

opt.linprog(c,a,b,d,e)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 0.52
              x: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00  0.000e+00
                   4.000e+01]
            nit: 2
          lower:  residual: [ 0.000e+00  6.000e+01  0.000e+00  0.000e+00
                              0.000e+00  4.000e+01]
                 marginals: [ 6.400e-03  0.000e+00  1.300e-03  3.000e-04
                              3.300e-03  0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00]
          eqlin:  residual: [ 0.000e+00]
                 marginals: [ 1.000e-03]
        ineqlin:  residual: [ 4.000e+00  0.000e+00  1.700e+00  1.000e-01]
                 marginals: [-0.000e+00 -7.000e-02 -0.000e+00 -0.000e+00]
 mip

**Resultados:**

Para que la mezcla sea barata y además cumpla con el estandar alimenticio establecido, se deben utilizar $60\;\mathrm{g}$ de carne y $40\;\mathrm{g}$ de gel con un valor total de $\mathrm{$}0.52$ por cada $100\;\mathrm{g}$

Además, se observa que esta mezcla tiene $12\;\mathrm{g}$ de proteínas, $6\;\mathrm{g}$ de grasa, $0.3\;\mathrm{g}$ de fibra y $0.3\;\mathrm{g}$ de sal.

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de tres plantas de generación para satisfacer la demanda diaria eléctrica en tres ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente.


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla.

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

In [1]:
import scipy.optimize as opt
import numpy as np
import matplotlib.pyplot as plt
import sympy as sym

In [15]:
#Función a minimizar de costos
c=np.array([5,2,7,3,6,6,6,1,2])

#Restricciones:
#Potencia máxima:
# x1 + x2 + x3 ≤ 80
# x4 + x5 + x6 ≤ 40
# x7 + x8 + x9 ≤ 60
#Demandas:
# x1 + x4 + x7 ≥ 70
# x2 + x5 + x8 ≥ 40
# x3 + x6 + x9 ≥ 70

a=np.array([[1,1,1,0,0,0,0,0,0],
            [0,0,0,1,1,1,0,0,0],
            [0,0,0,0,0,0,1,1,1],
            [-1,0,0,-1,0,0,-1,0,0],
            [0,-1,0,0,-1,0,0,-1,0],
            [0,0,-1,0,0,-1,0,0,-1]])
b=np.array([80,40,60,-70,-40,-70])

opt.linprog(c,a,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 540.0
              x: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01  0.000e+00
                   0.000e+00  0.000e+00  0.000e+00  6.000e+01]
            nit: 6
          lower:  residual: [ 3.000e+01  4.000e+01  1.000e+01  4.000e+01
                              0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+01]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              6.000e+00  1.000e+00  6.000e+00  4.000e+00
                              0.000e+00]
          upper:  residual: [       inf        inf        inf        inf
                                    inf        inf        inf        inf
                                    inf]
                 marginals: [ 0.000e+00  0.000e+00  0.000e+00  0.000e+00
                              0.000e+00  0.000e+00  0.000e+00  0.0

**Resultados:**

El costo mínimo para abastecer las tres ciudades es de $\mathrm{$} 540$. Guadalajara recibe 30 millones de $\mathrm{KW}$ de la planta 1 y 40 millones de $\mathrm{KW}$ de la planta 2. León recibe  40 millones de $\mathrm{KW}$ de la planta 1. Morelia recibe 10 millones de $\mathrm{KW}$ de la planta 1 y 60 millones de $\mathrm{KW}$ de la planta 3. Se satisfacen las necesidades de cada ciudad y no se excede la producción máxima de cada planta

### Fábrica de focos

Una fábrica produce focos incandesentes y los vende a 4.5 u.m., además de focos ahorradores que vende a 6 u.m. cada uno. La producción está limitada a 400 focos incandesentes y 300 focos ahorradores al día, además de no poder producir más de 500 focos en total.

Si la fábrica vende toda su producción, determina cuántos focos incandesentes y cuántos focos ahorradores debe producir para obtener los máximos ingresos posibles y cuáles serían éstos.

In [2]:
import scipy.optimize as opt
import numpy as np
import matplotlib.pyplot as plt
import sympy as sym

In [4]:
#Funcióna a maximizar
#F(i,a)= 4.5i + 6a

#Restricciones
# i ≤ 400
# a ≤ 300
# i + a ≤ 500

c=np.array([-4.5,-6])
a=np.array([[1,0],
            [0,1],
            [1,1]])
b=np.array([400,300,500])

opt.linprog(c,a,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: -2700.0
              x: [ 2.000e+02  3.000e+02]
            nit: 1
          lower:  residual: [ 2.000e+02  3.000e+02]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 2.000e+02  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -1.500e+00 -4.500e+00]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultados:**

Produciendo y vendiendo 200 focos incandesentes y 300 focos ahorradores, la ganancia máxima sería de $2,700\;\mathrm{u.m.}$. Se observa que se generan 200 focos incandesentes menos que lo máximo.

### Problema de compra al mayoreo
Un frutero necesita 16 cajas de naranjas, 5 de plátanos y 20 de manzanas. Dos mayoristas pueden suministrarle para satisfacer sus necesidades, pero sólo venden la fruta en contenedores completos. El mayorista A envía en cada contenedor 8 cajas de naranjas, 1 de plátanos y 2 de manzanas. El mayorista B envía en cada contenedor 2 cajas de naranjas, una de plátanos y 7 de manzanas. Sabiendo que cada contenedor del mayorista A tiene un costo de 150 u.m. y el contenedor del mayorista B de 300 u.m., calcular cuántos contenedores habrá de comprar a cada mayorista, con objeto de reducir al mínimo el costo de lo solicitado.

In [7]:
#Función a minimizar
#f(a,b) = 150a + 300b

#Restricciones
#Naranjas - 8a + 2b ≥ 16
#Plátanos - a + b ≥ 5
#Manzanas - 2a + 7b ≥ 20
c=np.array([150,300])
a=np.array([[-8,-2],
            [-1,-1],
            [-2,-7]])
b=np.array([-16,-5,-20])

opt.linprog(c,a,b)

        message: Optimization terminated successfully. (HiGHS Status 7: Optimal)
        success: True
         status: 0
            fun: 1050.0
              x: [ 3.000e+00  2.000e+00]
            nit: 3
          lower:  residual: [ 3.000e+00  2.000e+00]
                 marginals: [ 0.000e+00  0.000e+00]
          upper:  residual: [       inf        inf]
                 marginals: [ 0.000e+00  0.000e+00]
          eqlin:  residual: []
                 marginals: []
        ineqlin:  residual: [ 1.200e+01  0.000e+00  0.000e+00]
                 marginals: [-0.000e+00 -9.000e+01 -3.000e+01]
 mip_node_count: 0
 mip_dual_bound: 0.0
        mip_gap: 0.0

**Resultado:**

El frutero necesita sólo 3 contenedores A y 2 contenedores B, con un valor de $1050\;\mathrm{u.m.}$. Así paga lo mínimo, pero se pasa 12 naranjas del límite inferior con 28 naranjas.